### ISSUE TITLE
improving the real-time performance for superpoint network

### ISSUE THREAD

**Issue by shibowing** [rpautrat/SuperPoint#109]
﻿# improving the real-time performance for superpoint network

Hi, @rpautrat Thanks for your incredible work! 

Recently, I used your code to train superpoint network and tested it on Tesla K40c GPU combined with i7-3770 CPU. The real-time performance is not very promising and it takes 130ms for a single image with resolution **640*480**.  Through time line analysis, I found there are three most time-consuming parts:

 1. `pts = tf.to_float(tf.where(tf.greater_equal(prob, min_prob)))`  which can be find in box_nms function. It takes **59 ms!** The more detail can be found in this link.
(https://github.com/rpautrat/SuperPoint/blob/64e13e4e7341b8d32cf0339c7c4a520bea037d2a/superpoint/models/utils.py#L59)
 2. `indices = tf.image.non_max_suppression(boxes, scores, tf.shape(boxes)[0], iou)` which also can be find in box_nms function. It takes **28ms!** 
 3. I found that the descriptors "MEMCPYDtoH" takes **37ms!**
 4. Another weird problem, I found that the the vgg/conv1_2/conv/Conv2D takes **14ms!** compared with other Conv2D only takes **3ms!**

I want to improve the real-time performance of superpoint and run it on **TX2**. However, I do not know how to improve. Could you please give me some suggestions? 

In addition, I have used the ResizeBilinear instead of ResizeBicubic, which really saves some time. However, it still needs some other improvements to save time! I hope you can give me some suggestions. Thank you very much!        

This is timeline and bottom shows the most timecosting parts!
![Selection_001.png-130kB][1]
![Selection_002.png-163.1kB][2]
![Selection_003.png-228.1kB][3]
![Selection_004.png-129.1kB][4]

  [1]: http://static.zybuluo.com/machozhao/pop3ibq5kg6ijwos9m44dsgm/Selection_001.png
  [2]: http://static.zybuluo.com/machozhao/wmoqkhhfligpo6rmcb8mlcr5/Selection_002.png
  [3]: http://static.zybuluo.com/machozhao/6cbd4d589uh2b9huf9iu85in/Selection_003.png
  [4]: http://static.zybuluo.com/machozhao/odj6vsvzvp5y068j237q0kol/Selection_004.png

**Comment by rpautrat** [comment]
Hi!

Thanks a lot for sharing your findings! This code could indeed clearly be optimized further.

I am surprised though by the number 1 that you highlight. I don't see what would be so slow with this line of code... For the rest, there are probably improvements to be made. I wanted to have a look at it, but couldn't find for it so far.

You can try to use `spatial_nms` instead of `box_nms`, which is faster but doesn't handle ties. However since the output of the network is floating point values, I guess we can assume that there are never two points with exactly the same score predicted.

I am not sure where MEMCPYDtoH is called. I guess it is hidden in one of the tensorflow functions that copies data, maybe in `space_to_depth`?

A possible way to optimize the encoder and the convolutions would be to replace the VGG by a lighter network, like a MobineNet. This has already been done and offers a better efficiency while preserving almost the same performance.

**Comment by shibowing** [comment]
Thank you very much for your quick reply! I am very sorry for my late response.   According to your suggestions, I have modified the code. Although the real-time performance has been improved,right now only taking 55ms for network forward process on 640*480 images, the accuracy and stability are not very promising. Here is what I have modified and hope you can provide me some suggestions.
I have used mobilenetV2 to replace the vgg bakbone, which really improve the real-time performance. However, the accuracy and stability of the feature are not promising. It looks like the position of feature will move 2-3 pixels, even the thermal image only interval 3 frames.   May I ask , as your experience, what should I do if I want to improve the accuracy of feature point's location? I notice that the  detection_threshold=0.001 in coco.yaml and superpoint.yaml. Why this values so low? why not be 0.015(1/65)?
          
2. Also, I find that the mobilenet needs large number of nms (16 ). If I set nms =4 or 8,  the feaure position looks weird.  The true feature points will have 3 extra features located on the nearby pixel. The result can be found here. We set detection_threshold is 0.015. According to your experience, is that normal?
nms=4
![ms=4 mobilenet.png-456kB][5]
nms=8
![nms=8.png-447.1kB][6]
nms=16
![nms=16.png-439.7kB][7]

3.Although the real-time performance has been improved, compared to magicleap's pre-trained model, Our model is still slow(55ms for forward process). the pre-trained model only takes 33 ms for forward process.  Since I have already use mobilenetV2, our model theoretically should be faster than pre-trained model. Is that because of too many layers for mobilenet? The mobilenet structure can be found in here. I am very appreciated if you can share your specific mobilenet structure.

      spec=[
        op(slim.conv2d, stride=2, num_outputs=32, kernel_size=[3, 3]),
        op(ops.expanded_conv,
           expansion_size=expand_input(1, divisible_by=1),
           num_outputs=16),
        #op(ops.expanded_conv, stride=2, num_outputs=24),
        op(ops.expanded_conv, stride=1, num_outputs=24),
        #op(ops.expanded_conv, stride=2, num_outputs=32),
        op(ops.expanded_conv, stride=1, num_outputs=32),
        op(ops.expanded_conv, stride=1, num_outputs=32),
        op(ops.expanded_conv, stride=2, num_outputs=64),
        op(ops.expanded_conv, stride=1, num_outputs=64),
        op(ops.expanded_conv, stride=1, num_outputs=64),
        op(ops.expanded_conv, stride=1, num_outputs=64),
        op(ops.expanded_conv, stride=1, num_outputs=96),
        op(ops.expanded_conv, stride=1, num_outputs=96),
        op(ops.expanded_conv, stride=1, num_outputs=96),
        op(ops.expanded_conv, stride=2, num_outputs=128),
        op(ops.expanded_conv, stride=1, num_outputs=128),
        op(ops.expanded_conv, stride=1, num_outputs=128),
        op(ops.expanded_conv, stride=1, num_outputs=128),
        #op(slim.conv2d, stride=1, kernel_size=[1, 1], num_outputs=1280)
    ],

Also, I have used the spatial nms to replace box nms. The spatial nms is much worse than box nms. It seems it can not repress features robustly. Could you please recommend some other faster and robust NMS method?
I use this line to conduct spatial nms.
prob = tf.map_fn(lambda p: spatial_nms(p, config['nms']), prob) 
the reult of spatial nms. 
![spatial nms.png-461.6kB][8]
  [5]: http://static.zybuluo.com/machozhao/nrcfg3474zq0fbbgbszae0w2/ms=4%20mobilenet.png
  [6]: http://static.zybuluo.com/machozhao/2hd0qeeaneduccngpb5sbatx/nms=8.png
  [7]: http://static.zybuluo.com/machozhao/bfpht6gp9r3to87iip33zxv2/nms=16.png
  [8]: http://static.zybuluo.com/machozhao/2smtj8bhr4pi8q6tcaz16ikg/spatial%20nms.png

**Comment by rpautrat** [comment]
Hi,

1. & 2. I am working on a new version that solves this problem of multiple detections for a single interest point. This was due to a resize of the image during homographic adaptation, thus creating artifacts. This should also improve the accuracy of the points position. A detection threshold of 0.001 simply allowed me to get more predicted point, which improved training during the homography adaptation step.

3. I don't have a MobileNet architecture working for SuperPoint, I never did it actually. I was just mentioning that it was a possibility, but you will have to test various network structures I guess.
Have a look here: https://github.com/rpautrat/SuperPoint/issues/98, roborocklsm  was able to get ~80 FPS, so you can also ask him if he did other improvements.

4. Yes, that's why I used box_nms in the end. It was the best option (but slower) I found so far.

5. The detection threshold only plays a role in the homography adaptation, to determine the number of predictions that you want on a single image.

**Comment by roborocklsm** [comment]
@rpautrat 
Haha Yes I have achieved the inference speed ~80fps, but sadly it is on PC but not on Jetson TX2. 

@shibowing I guess you could use TensorRT to speed it up. For me, it helps a little bit.

**Comment by shibowing** [comment]
**Hi,  Many thanks for your kind advice! @rpautrat @roborocklsm** 

  1. I am still a little confused. if the H (homagraphy matrix) we caluclated is accurate,  why we have multiple detections for a single interest point? you mentioned that the resize of an image during homography is the reason, could you please explain more specifically? Do you mean there are some bugs in the fllowing  line? https://github.com/rpautrat/SuperPoint/blob/2962dde0687b36779291ea565c334487a89ce7fe/superpoint/models/homographies.py#L63
I do not understand the meaning "resize of the image during homographic adaptation improves the accuracy of the points position."
 


 2. Having carefurelly look, it seems like there are something wrong with homography matrix. 
 
   From this code, 
   `p_mat = tf.transpose(tf.stack(
        [[pts2[i][j] for i in range(4) for j in range(2)]], axis=0))
    homography = tf.transpose(tf.matrix_solve_ls(a_mat, p_mat, fast=True))`
    
According the equation below, it seems that the pts2[i][j] is [u2,v2,1] and pts1[i][j] is [u1,v1,1]. However, I find that our code seems inconsitent with this equation. It seems like 
`   H = sample_homography(shape, **config['homographies'])` **is not calcalating H. It is calculating H_inv. In the code the homography transform is using on pts2 not pts1. **However, according to the code above, I think the homography transform should be used on pts1.** ** 

Is there something wrong with my understanding? **Thanks for your explanation.^_^**
![image](https://user-images.githubusercontent.com/29483569/66613686-482b5f80-eb94-11e9-9ea8-20990219e68e.png)
![image (1)](https://user-images.githubusercontent.com/29483569/66613688-4cf01380-eb94-11e9-8e45-d5b92919c492.png)

**Comment by shibowing** [comment]
Hi, 
This is a kind reminder. Could you please reply above questions and give me some enlightenment (if you have free time)?
Thanks for your kind help!
^_^

**Comment by rpautrat** [comment]
Hi, sorry I had not time to look at it yet. I can at least tell you that:

1. The issue lies in this line rather: https://github.com/rpautrat/SuperPoint/blob/2962dde0687b36779291ea565c334487a89ce7fe/superpoint/models/homographies.py#L65
The resize on the probability spread the probability so that a single point is spread to multiple neighboring corners (and the spreading is large enough to be over the NMS distance). I will push the corrected version later (I want to correct something else before).
With this fix, the accuracy of the detection will be higher because the detection will be single and at the exact right position.

2. I didn't have time to carefully look at your comment, but I would be surprise if the homography computation is wrong in the code. I already spend a lot of time debugging it to get it to work and I am quite confident that it is doing the right thing now.
Note that the homography is used in various ways: the homography itself is used to warp the points, but it is the inverse which is used in the `tf.image.resize_images`. So sometimes it makes sense to have the inverse of the homography.
If you still have doubts, I encourage you to run simple tests to check if it is working as it should. Let me know if you come across something unexpected.

**Comment by shibowing** [comment]
I am sorry for the late reply! Thank you very much for your kind advice! O(∩_∩)O

**Comment by cumtchenchang** [comment]
@shibowing  Hi, I am replacing VGG with MobileNet to accurate the speed, have you realize? Can you give me more details? Thank you.
